include/login_functions.php

```diff
@@ -32,70 +35,88 @@ function setRememberMeCookie($userId)
+        // External authentication succeeded.
+        // Create the session array.
+        $sessionArray = array('login' => $login, 'password' => 'this is not a valid password', 'passwd' => $row['password'], 'ID' => session_id(), 'valid' => 1, 'loginid' => $userid);
         $_SESSION['cdash'] = $sessionArray;
-        pdo_free_result($result);
-        return true;                               // authentication succeeded
-    } elseif (md5($password) == $pass) {
-        // Authentication is successful.
-        if ($rememberme) {
-            setRememberMeCookie($user_array['id']);
+        return true;
+    } else {
+        $success = false;
+        if (password_verify($password, $user->Password)) {
+            $success = true;
+        } else if (md5($password) == $user->Password) {
+            // Re-hash this password using an algorithm that's more secure than md5.
+            // Do not attempt this before the database has been upgraded
+            // to accommodate the increased length of this field.
+            global $CDASH_DB_TYPE, $pdo;
+            $db_check = true;
+            if ($CDASH_DB_TYPE != 'pgsql') {
+                $table_name = qid('user');
+                $select = $pdo->query("SELECT password FROM $table_name LIMIT 1");
+                $meta = $select->getColumnMeta(0);
+                if ($meta['len'] < 255) {
+                    $db_check = false;
+                }
+            }
+            if ($db_check) {
+                $user->Password = password_hash($password, PASSWORD_DEFAULT);
+                $user->Save();
+            }
+            $success = true;
         }
 
-        session_name('CDash');
-        session_cache_limiter($SessionCachePolicy);
-        session_set_cookie_params($CDASH_COOKIE_EXPIRATION_TIME);
-        @ini_set('session.gc_maxlifetime', $CDASH_COOKIE_EXPIRATION_TIME + 600);
-        session_start();
+        if ($success) {
+            // Authentication is successful.
+            if ($rememberme) {
+                setRememberMeCookie($userid);
+            }
 
-        // create the session array
-        if (isset($_SESSION['cdash']['password'])) {
-            $password = $_SESSION['cdash']['password'];
-        }
-        $sessionArray = array('login' => $email, 'passwd' => $pass, 'ID' => session_id(), 'valid' => 1, 'loginid' => $user_array['id']);
-        $_SESSION['cdash'] = $sessionArray;
+            session_name('CDash');
+            session_cache_limiter($SessionCachePolicy);
+            session_set_cookie_params($CDASH_COOKIE_EXPIRATION_TIME);
+            @ini_set('session.gc_maxlifetime', $CDASH_COOKIE_EXPIRATION_TIME + 600);
+            session_start();
 
-        checkForExpiredPassword();
-        clearUnsuccessfulAttempts($user_array['id']);
-        return true;
+            // Create the session array.
+            $sessionArray = array('login' => $email, 'passwd' => $user->Password, 'ID' => session_id(), 'valid' => 1, 'loginid' => $userid);
```

Comment:
Here's where this is used (link to master):
https://github.com/Kitware/CDash/blob/master/include/login_functions.php#L366

If I'm reading it right, we store the hash of the password in the session to help make sure that the session is valid.